### ISSUE TITLE
incorrect macro expansion for R6RS define-record-type

### ISSUE THREAD

**Issue by larceny-trac-import** [larcenists/larceny#633]
**Reported by: will  on Tue Apr  7 19:40:28 2009**
As reported by Derick Eddington.  (Will believes this bug report
refers to R6RS ```define-record-type```, not to SRFI 9 or ERR5RS
```define-record-type```, although they may have similar problems.)

```foo``` is macro-introduced (which means the exception for the
last expression below is correct) but the other identifiers are
not, so they should retain their original lexical context, which
means the exceptions for the ```foo-x``` and ```foo-x-set!```
expressions below are incorrect because those identifiers should
be bound.

```
Larceny v0.97a4 (alpha test) (Mar 28 2009 23:47:58, precise:Linux:unified)
larceny.heap, built on Sat Mar 28 23:49:25 PDT 2009
ERR5RS mode (no libraries have been imported)

> (import (rnrs))
Autoloading (rnrs)
Autoloading (rnrs enums)
Autoloading (rnrs lists)
Autoloading (rnrs syntax-case)
Autoloading (rnrs conditions)
Autoloading (err5rs records procedural)
Autoloading (rnrs exceptions)
Autoloading (rnrs hashtables)
Autoloading (rnrs arithmetic bitwise)
Autoloading (rnrs programs)
Autoloading (rnrs files)
Autoloading (rnrs io ports)
Autoloading (larceny deprecated)
Autoloading (rnrs records syntactic)
Autoloading (rnrs records procedural)
Autoloading (rnrs control)
Autoloading (rnrs sorting)
Autoloading (rnrs bytevectors)
Autoloading (rnrs unicode)

> (define-syntax def-foo
    (syntax-rules ()
      ((_ make pred (field accessor setter) ...)
       (define-record-type (foo make pred)
         (fields (mutable field accessor setter) ...)))))

> (def-foo make-foo foo? (x foo-x foo-x-set!))

> foo-x


Error: Undefined global variable "foo-x".
Entering debugger; type "?" for help.
debug> q

> foo-x-set!


Error: Undefined global variable "foo-x-set!".
Entering debugger; type "?" for help.
debug> q

> make-foo
#<PROCEDURE>

> foo?
#<PROCEDURE condition?>

> foo


Error: Undefined global variable "foo".
```

**Comment by WillClinger** [comment]
Yes, SRFI 99 has a similar problem, only worse:
```
Larceny v0.98 "General Ripper" (Mar  7 2015 01:06:26, precise:Linux:unified)
larceny.heap, built on Sat Mar  7 01:06:50 EST 2015

> (import (srfi 99))

> (define-syntax def-foo
    (syntax-rules ()
      ((_ make pred (field accessor setter) ...)
       (define-record-type foo
         make
      	 pred
      	 (field accessor setter) ...))))

> (def-foo make-foo foo? (x foo-x foo-x-set!))
foo-x-set!

> foo-x

Error: unhandled condition:
Compound condition has these components: 
#<record &assertion>
#<record &message>
    message : "?: undefined global variable: foo-x \n"

Entering debugger; type "?" for help.
debug> q

> foo-x-set!

Error: unhandled condition:
Compound condition has these components: 
#<record &assertion>
#<record &message>
    message : "?: undefined global variable: foo-x-set! \n"

Entering debugger; type "?" for help.
debug> q

> make-foo

Error: unhandled condition:
Compound condition has these components: 
#<record &assertion>
#<record &message>
    message : "?: undefined global variable: make-foo \n"

Entering debugger; type "?" for help.
debug> q

> foo?

Error: unhandled condition:
Compound condition has these components: 
#<record &assertion>
#<record &message>
    message : "?: undefined global variable: foo? \n"

Entering debugger; type "?" for help.
debug> q

> foo

Error: unhandled condition:
Compound condition has these components: 
#<record &assertion>
#<record &message>
    message : "?: undefined global variable: foo \n"

Entering debugger; type "?" for help.
debug> a
```

**Comment by WillClinger** [comment]
Fixed by changeset 25ae9d0e0bf1d7f6e6c76dd9fa79c607466d5f0b

**Comment by WillClinger** [comment]
Oops, that introduced a new bug that was then fixed by changeset 86989e6c53fe83ad9e941706c48c3458a6061ddb
